toolbox/githubctl/main.go

```diff
@@ -79,167 +80,97 @@ func fastForward(repo, baseBranch, refSHA *string) error {
 	return githubClnt.FastForward(*repo, *baseBranch, *refSHA)
 }
 
-func processIstioVersion(content *string) map[string]string {
-	kv := make(map[string]string)
-	lines := strings.Split(*content, "\n")
-	for _, line := range lines {
-		if idx := strings.Index(line, "="); idx != -1 {
-			key := line[len(export):idx]
-			value := strings.Trim(line[idx+1:], "\"")
-			kv[key] = value
-		}
-	}
-	return kv
-}
-
-func getReleaseTag() (string, error) {
-	u.AssertNotEmpty("base_branch", baseBranch)
-	releaseTag, err := githubClnt.GetFileContent(istioRepo, *baseBranch, releaseTagFile)
-	if err != nil {
-		return "", err
-	}
-	return strings.Trim(releaseTag, "\n"), nil
+type task struct {
+	job       string
+	runNumber int
 }
 
-// TagIstioDepsForRelease creates release tag on each dependent repo of istio
-func TagIstioDepsForRelease() error {
-	u.AssertNotEmpty("base_branch", baseBranch)
-	log.Printf("Fetching and processing istio.VERSION\n")
-	istioVersion, err := githubClnt.GetFileContent(istioRepo, *baseBranch, istioVersionFile)
-	if err != nil {
-		return err
-	}
-	kv := processIstioVersion(&istioVersion)
-	pickledDeps, err := githubClnt.GetFileContent(istioRepo, *baseBranch, istioDepsFile)
-	if err != nil {
-		return err
-	}
-	deps, err := u.DeserializeDepsFromString(pickledDeps)
-	if err != nil {
-		return err
-	}
-	log.Printf("Fetching release tag\n")
-	releaseTag, err := getReleaseTag()
-	if err != nil {
-		return err
-	}
-	releaseMsg := "Istio Release " + releaseTag
-	for _, dep := range deps {
-		// use sha directly read from istio.VERSION in case updateVersion.sh was run
-		// manually without also updating istio.deps
-		log.Printf("Creating annotated tag [%s] on %s\n", releaseTag, dep.RepoName)
-		ref, exists := kv[dep.Name]
-		if !exists {
-			return fmt.Errorf("ill-defined %s: unable to find %s", istioVersionFile, dep.Name)
-		}
-		// make sure ref is a SHA, special case where previous release is used in this release
-		if u.ReleaseTagRegex.MatchString(ref) {
-			ref, err = githubClnt.GetTagCommitSHA(dep.RepoName, ref)
-			if err != nil {
-				return err
-			}
-		}
-		if err := githubClnt.CreateAnnotatedTag(
-			dep.RepoName, releaseTag, ref, releaseMsg); err != nil {
-			if strings.Contains(err.Error(), "Reference already exists") {
-				log.Printf("Tag [%s] already exists on %s\n", releaseTag, dep.RepoName)
-				prevTagSHA, err := githubClnt.GetTagCommitSHA(dep.RepoName, releaseTag)
-				if err != nil {
-					return err
+// preprocessProwResults downloads the most recent prow results up to maxRunDepth
+// then returns a two-level map job -> sha -> passed (true) or failed (false)
+func preprocessProwResults() map[string]map[string]bool {
+	log.Printf("Start preprocessing prow results")
+	prowAccessor := s.NewProwAccessor(
+		prowProject,
+		prowZone,
+		gubernatorURL,
+		gcsBucket,
+		u.NewGCSClient(gcsBucket))
+	cache := make(map[string]map[string]bool)
+	tasks := make(chan task)
```

Comment:
Add size, make the task a pointer, name the channel as tasksCh

tasksCh := make(chan *task, *maxConcurrentRequests)

Reply:
done.

Passing pointers instead reduces the total execution time by half. Awesome.